decidim-core/app/controllers/decidim/links_controller.rb

```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+module Decidim
+  class LinksController < Decidim::ApplicationController
+    helper Decidim::ExternalDomainHelper
+
+    rescue_from Decidim::InvalidUrlError, with: :invalid_url
+
+    def new
```

Comment:
We should check that the `external_url` parameter exists because the helper will still throw an unexpect exception when the parameter value is set to `nil`. The helper tries to run the match against `nil`.

You can try this by requesting the page without any parameters: `http://localhost:3000/link`

Actually, now thinking twice about this, I think it would be safer to run the regexps and checks for the URL in the controller because now the functionality of the view depends on the helper method. If that method call is removed from the view, it will break.

Let's add a `before_action` that does the necessary parsing for the URL (do the necessary changes in the helper as well after this):

```ruby
  before_action :parse_url

  private

  def parse_url
    # Move the error class under the controller from under the helper
    raise Decidim::InvalidUrlError unless external_url

    parts = external_url.match %r{^(([a-z]+):)?//([^/]+)(/.*)?$}
    raise Decidim::InvalidUrlError unless parts

    # Use these in the view
    @url_parts = {
      protocol: parts[1],
      domain: parts[3],
      path: parts[4]
    }
  end

  def external_url
    @external_url ||= params[:external_url]
  end
```